core/src/main/java/io/questdb/griffin/engine/functions/analytic/RankFunctionFactory.java

```diff
@@ -0,0 +1,182 @@
+/*******************************************************************************
+ *     ___                  _   ____  ____
+ *    / _ \ _   _  ___  ___| |_|  _ \| __ )
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2022 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.griffin.engine.functions.analytic;
+
+import io.questdb.cairo.*;
+import io.questdb.cairo.map.*;
+import io.questdb.cairo.sql.Record;
+import io.questdb.cairo.sql.*;
+import io.questdb.griffin.FunctionFactory;
+import io.questdb.griffin.SqlException;
+import io.questdb.griffin.SqlExecutionContext;
+import io.questdb.griffin.engine.RecordComparator;
+import io.questdb.griffin.engine.analytic.AnalyticContext;
+import io.questdb.griffin.engine.analytic.AnalyticFunction;
+import io.questdb.griffin.engine.functions.LongFunction;
+import io.questdb.griffin.engine.orderby.RecordComparatorCompiler;
+import io.questdb.std.*;
+
+public class RankFunctionFactory implements FunctionFactory {
+
+    @Override
+    public String getSignature() {
+        return "rank()";
+    }
+
+    @Override
+    public Function newInstance(int position, ObjList<Function> args, IntList argPositions, CairoConfiguration configuration, SqlExecutionContext sqlExecutionContext) {
+        final AnalyticContext analyticContext = sqlExecutionContext.getAnalyticContext();
+
+        if (analyticContext.getPartitionByRecord() != null) {
+            ArrayColumnTypes arrayColumnTypes = new ArrayColumnTypes();
+            arrayColumnTypes.add(ColumnType.LONG); // max index
+            arrayColumnTypes.add(ColumnType.LONG); // current index
+            arrayColumnTypes.add(ColumnType.LONG); // offset
+            Map map = MapFactory.createMap(configuration, analyticContext.getPartitionByKeyTypes(), arrayColumnTypes);
+            return new RankFunction(map, analyticContext.getPartitionByRecord(), analyticContext.getPartitionBySink());
+        }
+        return new SequenceRankFunction();
```

Comment:
Is it correct to use `SequenceRankFunction` in the `rank() OVER (ORDER BY col_a, col_b)` case? Looks like we need one more function to handle this case.

`SequenceRankFunction` should be used only in the `rank() OVER ()` case. Moreover, as it always returns `1`, we should use `LongConstant.newInstance(1)` here instead of defining a dedicated `SequenceRankFunction` class.

Also, we need some tests for both of these cases.

Reply:
Function `row_number()` not support  over clause which has partition by, should we supports `rank() OVER (ORDER BY col_a, col_b)` ?

Reply:
Perhaps function row_number() not support  partition by in over clause is an issue #2604 ？

Reply:
> Function `row_number()` not support over clause which has partition by, should we supports `rank() OVER (ORDER BY col_a, col_b)` ?

Would be nice to support all cases in `rank()`. `row_number()` can be fixed separately or, if you'd like to go an extra mile, in this PR.

Reply:
i will fix `row_number()` issue in new PR.